presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -0,0 +1,603 @@
+
+                            if (!matchingSubfields.isEmpty()) {
+                                found = true;
+                                matchingSubfields.stream()
+                                        .map(Subfield::getPath)
+                                        .map(path -> new Subfield(container.getName(), ImmutableList.<Subfield.PathElement>builder()
+                                                .add(allSubscripts())
+                                                .addAll(path)
+                                                .build()))
+                                        .forEach(newSubfields::add);
+                            }
+                        }
+                    }
+                }
+                if (!found) {
+                    context.get().variables.add(container);
+                }
+            }
+            context.get().subfields.addAll(newSubfields.build());
+
+            return context.defaultRewrite(node, context.get());
+        }
+
+        @Override
+        public PlanNode visitWindow(WindowNode node, RewriteContext<Context> context)
+        {
+            context.get().variables.addAll(node.getSpecification().getPartitionBy());
+
+            node.getSpecification().getOrderingScheme()
+                    .map(OrderingScheme::getOrderBy)
+                    .ifPresent(context.get().variables::addAll);
+
+            node.getWindowFunctions().values().stream()
+                    .map(WindowNode.Function::getFunctionCall)
+                    .map(CallExpression::getArguments)
+                    .flatMap(List::stream)
+                    .map(OriginalExpressionUtils::castToExpression)
+                    .forEach(expression -> subfieldExtractor.process(expression, context.get()));
+
+            node.getWindowFunctions().values().stream()
+                    .map(WindowNode.Function::getFrame)
+                    .map(WindowNode.Frame::getStartValue)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .forEach(context.get().variables::add);
+
+            node.getWindowFunctions().values().stream()
+                    .map(WindowNode.Function::getFrame)
+                    .map(WindowNode.Frame::getEndValue)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .forEach(context.get().variables::add);
+
+            return context.defaultRewrite(node, context.get());
+        }
+
+        private boolean isRowType(VariableReferenceExpression variable)
+        {
+            return variable.getType() instanceof ArrayType && ((ArrayType) variable.getType()).getElementType() instanceof RowType;
+        }
+
+        private static boolean prefixExists(Subfield subfieldPath, Collection<Subfield> subfieldPaths)
+        {
+            return subfieldPaths.stream()
+                    .filter(path -> path.isPrefix(subfieldPath))
```

Comment:
nit: IntelliJ says this this can be replaced with `anyMatch`:

```java
            return subfieldPaths.stream()
                    .anyMatch(path -> path.isPrefix(subfieldPath));
```